### PR TITLE
EKF: Improve covariance prediction stability

### PR DESCRIPTION
**Pull request by priseborough**
Eliminates collapse of vertical velocity state variance due to rounding errors that can occur under some operating conditions.

See issue  https://github.com/PX4/ecl/issues/771 for discussion and testing against stable release version.

TODO: testing of this version

### CODE REVIEW

File: EKF/covariance.cpp [hunk 1]
```diff
@@ -734,12 +735,12 @@ void Ekf::fixCovarianceErrors(bool force_symmetry)
 
 	for (int i = 4; i <= 6; i++) {
 		// NED velocity states
-		P(i,i) = math::constrain(P(i,i), 0.0f, P_lim[1]);
+		P(i,i) = math::constrain(P(i,i), 1E-6f P_lim[1]);
```
Comment (jkflying) [inline EKF/covariance.cpp:738]: Missing a comma here
```suggestion
		P(i,i) = math::constrain(P(i,i), 1E-6f, P_lim[1]);
```

**Comment by dusan19** [comment]
Would it be benefitial if the filter update period was increased to 12ms to keep it as a multiple of the incoming  imu samples?

**Comment by priseborough** [comment]
> Would it be benefitial if the filter update period was increased to 12ms to keep it as a multiple of the incoming imu samples?

We can have significant dt jitter on IMU sampling depending on the board, so there is no benefit  in selecting an integer multiple given the way the imu_down_sampler.cpp is adjusting the time cutoff to maintain the average rate at the target.

**Comment by priseborough** [comment]
We could move to a fixed down sampling ratio, eg 3:1, but it would rely on the published IMU data dt having low jitter.

**Comment by priseborough** [comment]
For example the BMI088 sensor is sampled at 800Hz which does not downsample to 250Hz with even time steps.

**Comment by priseborough** [comment]
If we went to 12, then maybe we could use the instantaneous IMU dt instead of the average dt inside the covariance prediction as the likelihood of a 1 frame IMU sample being consumed then becomes small, but I haven't tested that option. Using 12 instead of 10 will also increase timing jitter for other sensors wrt the fusion time horizon.

**Comment by priseborough** [comment]
I'm working through the CI test failures one by one adjusting the unit tests to get them to pass. Can someone else have a look at the  imuSamplingAtMultipleRates test. What it is doing doesn't make sense to me, but I may be misunderstanding it.

**Comment by priseborough** [comment]
The only tests failing now are:

EkfExternalVisionTest.visionVelocityReset
EkfExternalVisionTest.visionVelocityResetWithAlignment
 
It's not obvious how this change is impacting these cases, so further investigation is required.

**Comment by jkflying** [comment]
@kamilritz do you have time to take a quick look at why those tests are failing?

**Comment by priseborough** [comment]
I've got them all passing now.

**Comment by kamilritz** [comment]
Cool, Another option is to increase the time before the reset and capture the state of after the reset a bit earlier, like this: https://github.com/kamilritz/ecl/commit/eea01467b75d5f4f822cdc950e8ec58398729dc0
I would prefer this option since with this check we want to test if the reset happens to the correct value and not if we converge towards the right value.
@priseborough Are you able to update the change indication locally? Otherwise you could copy-replace it from here: https://github.com/kamilritz/ecl/commit/528a235afb190a20c499056a445294b49a8f9d1c

**Comment by dagar** [comment]
> We can have significant dt jitter on IMU sampling depending on the board, so there is no benefit in selecting an integer multiple given the way the imu_down_sampler.cpp is adjusting the time cutoff to maintain the average rate at the target.

It's significantly better with the new Invensense drivers in master (and the v1.11 release). Currently that's the primary IMU in the vast majority of usage. Note these still have the hard coded 4000 us integration period, but it's handled as an integer number of samples. We could easily change this to 2500 us.

> We could move to a fixed down sampling ratio, eg 3:1, but it would rely on the published IMU data dt having low jitter.

I would like to consider that after this release. I don't see why we can't optimize the system holistically, including accommodating these fundamental differences between sensors.

**Comment by dagar** [comment]
For testing I've created https://github.com/PX4/Firmware/pull/14676 that pulls in this ECL branch and also changes the default integration period.

**Comment by priseborough** [comment]
> @priseborough Are you able to update the change indication locally? Otherwise you could copy-replace it from here: [kamilritz@528a235](https://github.com/kamilritz/ecl/commit/528a235afb190a20c499056a445294b49a8f9d1c)

Not with the current level of documentation for the change indication. I know you showed me once before, but I thought you were going to write a readme that specifies the steps required to generate a new file

**Comment by priseborough** [comment]
@kamilritz I pulled in your suggested test changes, but had to resolve a conflict and there are still there are issues with unit  tests failing after the rebase.

**Comment by kamilritz** [comment]
@priseborough Great, that you could get those tests to pass. These tests seem to be really time sensitive. Maybe it is due to the fact that we compare the output velocities and not the velocities of the delayed state. I will look in to this.

**Comment by dusan19** [comment]
I tested this changes on a log that was failing due to covariance prediction instability that I recently recorded (sorry i didnt do it sooner before this PR has been merged). The results are posted here: https://github.com/PX4/ecl/issues/771#issuecomment-618596373
As the plots show, this changes don't solve the problem but might make it slightly better, although the drone would still perform the fallback landing due to invalid position estimate (and crash)

**Pull request merged**